fuel-relayer/src/validators.rs

```diff
@@ -26,16 +26,94 @@ impl Validators {
         self.set = db.get_validators().await;
     }
 
-    /// Get validator set
+    /// Get validator set for da_height.
     pub async fn get(
-        &mut self,
+        &self,
         da_height: DaBlockHeight,
+        db: &mut dyn RelayerDb,
     ) -> Option<HashMap<Address, (u64, Option<Address>)>> {
-        // TODO apply down drift https://github.com/FuelLabs/fuel-core/issues/365
-        if self.da_height == da_height {
-            return Some(self.set.clone());
+        match self.da_height.cmp(&da_height) {
+            std::cmp::Ordering::Less => {
+                // We request validator set that we still didnt finalized or know about.
+                // Probably there is eth client sync problem
+                error!(
+                    "current height {} is less then requested validator set height: {da_height}",
+                    self.da_height
+                );
+                return None;
+            }
+            std::cmp::Ordering::Equal => {
+                // unusual but do nothing
+                info!("Get last finalized set height: {da_height}");
+                let set = self
+                    .set
+                    .iter()
+                    .filter(|(_, (stake, consensus_key))| *stake != 0 && consensus_key.is_some())
+                    .map(|(k, v)| (*k, *v))
+                    .collect();
+                return Some(set);
+            }
+            std::cmp::Ordering::Greater => {
+                // slightly drift to past
+            }
+        }
+
+        let mut validators = self.set.clone();
+        // get staking diffs to revert from our current set.
+        let diffs = db
+            .get_staking_diffs(da_height + 1, Some(self.da_height))
+            .await;
+
+        for (diff_height, diff) in diffs.into_iter().rev() {
+            // update consensus_key
+            for (
+                validator,
+                ValidatorDiff {
+                    previous_consensus_key,
+                    ..
+                },
+            ) in diff.validators
+            {
+                if let Some((_, consensus_key)) = validators.get_mut(&validator) {
+                    *consensus_key = previous_consensus_key;
+                } else {
+                    panic!("Validator should be present when reverting diff");
+                }
+            }
+
+            // for every delegates, cache it and if it is not in cache query db's delegates_index for earlier delegate set.
+            for (delegator, delegation) in diff.delegations.into_iter() {
```

Comment:
[nit] unneeded cast to iter in for loop
```suggestion
            for (delegator, delegation) in diff.delegations {
```